### ISSUE TITLE
Autopicking depends on order in input micrographs.star

### ISSUE THREAD

**Issue by bforsbe** [3dem/relion#548]
I'm seeing major differences in autopicking results where there should be no differences. 

The runs conducted are: 
 - Run1: Helical autopick with 50 mics
 - Run2: same as Run1, but one mic excluded (i.e. input file has 49 mics instead of 50)
 - Run3: same as Run2 (repeat of Run2) 
 - Run4: same as Run2, but with 3 MPIs instead of 4 (see below for full command)
 - Run5: same as Run2, but randomized order of mics in input star file. 
 - Run6: same as Run1, but not using MPI
 - Run7: same as Run2, but not using MPI

There are differences in the number of picked particles for all mics, but I illustrate how large the difference is for just one: 

Picked particles: 
 - Run1: 360
 - Run2: 160
 - Run3: 160 (all mics same count as Run2)
 - Run4: 160 (all mics same count as Run2)
 - Run5: 318
 - Run6: 360 (all mics same count as Run1)
 - Run7: 160 (all mics same count as Run2)

**Environment:**
 - gcc: 7.3
 - CUDA: 10.1.2
 - MPI runtime: 4.0.0
 - RELION version 3.1-beta-74beeb 
 - GPU: 4 x GTX1080

**Job options:**
 - Type of job: Autopick
 - Number of MPI processes: 4
 - Number of threads: 1
 - Full command:

```
relion_autopick_mpi --i <file> --odir AutoPick/jobxxx/ --pickname autopick --ref <file> --invert  --ctf  --ang 3 --shrink 0.8 --lowpass 10 --angpix_ref 3.3125 --particle_diameter 300 --threshold 1 --min_distance 62.8 --max_stddev_noise 1.1 --helix --helical_tube_outer_diameter 200 --helical_tube_kappa_max 0.8 --helical_tube_length_min 300 --gpu  --pipeline_control AutoPick/jobxxx/
```
I can provide screenshots and data for debugging, but as this is unpublished data I refrain from posting it here. I get a warning about splitting the computatio into batches due to lowpass 10Å and shrink 0.8:
```
With the current settings the GPU memory is imposing a soft limit on your performace,
since one or more micrographs has to use (at least 4) batches of orientations to 
achieve the total requested 120 orientations. Consider using
	 higher --ang
	 harder --shrink
	 higher --lowpass with --shrink 0
```
But this *really* looks like it's dependent on the order of the mics in the input file rather than how they are processed.I'm surprised that the number of MPIs don't make a difference though....

**Comment by biochem-fan** [comment]
Interesting. Does this happen on the CPU version? Can you write the FOM maps and see if it changes? Note that AutoPick uses random numbers so the result will not be exactly the same. But this difference is too large.

**Comment by bforsbe** [comment]
Didnt think about the random seed. I'll try the picking on cpu and maybe back the code to use the sum of a mic as its random seed to get rid of the dependence of order in the input for this factor at least.

**Comment by bforsbe** [comment]
The CPU run is taking it's sweet time but I can say that the random seed makes a surprising difference. Setting ```--random_seed 3412387``` makes the picking much better for some mics, and much worse for others. Crazy that it should have such influence.  I've sent you an email with a screenshot, since this is unpublished data.

**Comment by bforsbe** [comment]
CPU behaves very similar to GPU, i.e. the particle numbers are the same +/- 1% compared to the GPU-run, so I guess mostly it's the default precision and rounding errors that constitute the difference. I would conclude that the rand_seed or other is responsible rather than being GPU-specific.

**Comment by biochem-fan** [comment]
Does this affect single particles as well?

**Comment by biochem-fan** [comment]
I discussed this with @scheres. I will write to you this afternoon.

**Comment by bforsbe** [comment]
I tried to autopick the same image twice as separate entries in the input mic.star and I still got wildly different picking numbers. If I set the randomization to initialize the same for each mic by hacking the code here 

https://github.com/3dem/relion/blob/ce28d249ecb06db7a2a121f2e784accb8c99ff2c/src/acc/cuda/cuda_autopicker.cu#L293

I get reproducible results, but it varies with random seed as described before. By turning off the randomization here; 
https://github.com/3dem/relion/blob/ce28d249ecb06db7a2a121f2e784accb8c99ff2c/src/acc/cuda/cuda_autopicker.cu#L862

I get rubbish output, but identical CCF, so I'm willing to wager that is where the sensitivity lies. I say sensitivity because I'm not sure it's an issue, but probably an instability in the code for certain samples and modes of picking. I guess it could be due to false assumptions regarding normalization of mics or references, but I don't really understand the math in this part of the code.
